Automattic/woocommerce-payments
# Refactor the SQL query modifications in Analytics

Fixes #4632

## Changes proposed in this Pull Request

Rework `WCPay\MultiCurrency\Analytics` in order to support Custom Order Tables.

### What does the class do?

@dmallory42 described what the class does in https://github.com/Automattic/woocommerce-payments/issues/4632#issuecomment-1232681417. Here is the summary:

- The class modifies the queries, made by WooCommerce Analytics to adjust for order's currencies.
- WC Analytics has custom queries, so neither `get_posts` nor `wc_get_orders` are used or modified here.
- The additional data, needed to adjust the queries, is stored as order meta.

### Analytics and Custom Order Tables

I didn't know whether Analytics should work with COT, so I asked in a couple of places in Slack without getting a clear answer:

1. p1662026947873699-slack-C03TN23TZB5
2. p1662026407137989-slack-C9QDV4TL4

All orders (custom order table and not) on my local instance were already deleted so I could have a clean slate, and begin fixing the issue with order data contained only in custom order tables, to make sure that Analytics are not relying on missing data. This could have been a bad decision.

After creating a few orders, and experimenting with [Analytics' settings](http://localhost:8082/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Fsettings), mainly `Import Historical Data`, I could not get those orders to show up in Analytics.

My conclusion is that the only way to test this (PR) right now is to:

1. Enable custom order tables, and choose to use WooCommerce Orders tables, and to keep synchronization on in [custom data stores section](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=advanced&section=custom_data_stores).
2. Create a few orders, which will store the data both in custom order tables, as well as posts.
3. Modify the code to support reading from both datastores, **while ensuring that in COT mode we are not accessing `wp_posts` and `wp_postmeta` in any way**, and vice-versa. See the next point.

### The changes

In _What does the class do?_ I already mentioned that all of the required additional data is stored as order meta. After looking at the class, I realized that it's job is primarily to `LEFT JOIN` the `wp_postmeta` table a few times, and to modify the rest of the query accordingly. <-- @dmallory42 could you validate this assumption?

Within this PR I use `OrderUtil::custom_orders_table_usage_is_enabled()`, and if enabled, I switch from joining `wp_postmeta` to joining `wp_wc_orders_meta` while using `order_id` as a key, instead of `post_id`. Both meta tables have the same structure, except for the name of the foreign key column, so this work.

As an additional modification, I changed the aliases of those tables within the query to just use `meta` instead of `postmeta` for clarity.

## Testing instructions

1. Set up COT to use the orders table, and keep sync on.
2. Make sure the posts and orders tables are synchronized (WooCommerce > Status > Action Scheduler > Pending > Run all jobs).
3. Make sure you have Multi-Currency set up, and at least one order, which was paid in a non-default currency.
4. Open an Analytics page, any of them should work, but I prefer the overview.
5. Clone the tab (in order to keep this one for reference).
6. **Only then check out this branch**
7. Refresh the cloned tab and **ensure that nothing changes**.
8. Inspect the performed SQL query/queries, and make sure that all of them join the `wp_wc_orders_meta` table, and `wp_postmeta` is nowehere in sight.

### How to inspect the performed SQL queries?

This one is a bit tricky:

- I like using [Query Monitor](https://wordpress.org/plugins/query-monitor/), but because all of the data is loaded through AJAX calls, those queries do not appear there.
- My initial choice was to use XDebug, and pause somewhere in the code to inspect the SQL, but somehow it got truncated all of the time...

I ended up modifying `Automattic\WooCommerce\Admin\API\Reports\Orders\Stats\DataStore::get_data()` (source [here](https://github.com/woocommerce/woocommerce/blob/cot-main/plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php)). I replaced this piece of code:

https://github.com/woocommerce/woocommerce/blob/9b3aee30f3185455af09db7c172eabf7ac49b644/plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php#L330-L333

with:

```php
$query_statement = $this->total_query->get_query_statement();
$file = fopen( '/var/www/html/analytics-query.sql', 'a' );
fwrite( $file, $query_statement . "\n\n" );
fclose( $file );
$totals = $wpdb->get_results(
	$query_statement,
	ARRAY_A
); // phpcs:ignore cache ok, DB call ok, unprepared SQL ok.
```

Then I reloaded the page, and checked out `docker/wordpress/analytics-query.sql` in my editor. It might also help to empty the file before reloading the page.

-------------------
